### ISSUE TITLE
Prevent buggy simulation behavior due to lack of mesh preparedness

### ISSUE THREAD

**Issue by smharper** [idaholab/moose#15944]
The meshgenerator system is currently missing calls to `prepare_for_use` and this can result in segfaults.

For example, this example with `GeneratedMeshGenerator` causes a segfault:
```
[Mesh]
  [./gmg]
    type = GeneratedMeshGenerator
    dim = 1
    nx = 1
    ny = 0
    nz = 0
    elem_type = EDGE2
  []

  [./extrude]
    type = FancyExtruderGenerator
    input = gmg
    heights = '1'
    num_layers = '1'
    direction = '0 1 0'
  []
[]

[Variables]
  [./u]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
  [../]
[]

[BCs]
  [./bottom]
    type = DirichletBC
    variable = u
    boundary = '0'
    value = 0
  []
  [./top]
    type = DirichletBC
    variable = u
    boundary = '1'
    value = 1
  [../]
[]

[Executioner]
  type = Steady
  solve_type = 'PJFNK'
[]

[Outputs]
  exodus = true
[]
```

Related issues: #15936 #15823

@friedmud and I spoke offline about this issue. He convinced me that we can't simply call `prepare_for_use` after each call to a `MeshGenerator::generate()` function. There needs to be some mechanism by which meshgenerators like `FancyExtruderGenerator` advertise that their output mesh is unprepared and other meshgenerators like `SideSetsBetweenSubdomainsGenerator` advertise that they need a prepared mesh.

**Comment by friedmud** [comment]
This should be fixed by #15945

**Comment by lindsayad** [comment]
I'm adding this to the Pronghorn/CRAB project since we're using `FancyExtruderGenerator` in FV input files

**Comment by lindsayad** [comment]
As @friedmud said the long-term fix is #15945. However, in the meanwhile we can address this by tracking `is_prepared` and then making calls to a new API called `set_isnt_prepared` or something like that as described in https://github.com/idaholab/moose/issues/15945#issuecomment-708649237. Then we can remove all the `prepare` jibber-jabber in `MooseMesh`. This is the ticket I'll close when when we have this short-term fix in.
